proposals/call-tags/Overview.md

```diff
@@ -4,183 +4,254 @@
-
-The `func_switch` extension means that the behavior of a `funcref` can depend on the specific call tag used (beyond just matching versus trapping).
-Using this, we can support an efficient non-JITing implementation of interface-method dispatch (and other lesser-known forms of dynamic dispatch) for many languages with multiple inheritance of interfaces.
-For context, one way to implement interface-method dispatch is to have every v-table have an array of, say, 19 slots, and to assign to every interface method some slot number.
-Unfortunately, it is possible that an object implements multiple interface methods with the same slot number.
-So when a interface-method call is made, in addition to the arguments the caller pushes onto the stack the identifier of the interface method (i.e. its call tag) and then calls the function in the matching slot.
-That function then switches on that identifier and redirect to the appropriate implementation, just like a `func_switch`.
-See [Efficient implementation of Java interfaces: Invokeinterface considered harmless](https://dl.acm.org/doi/10.1145/504282.504291) for more information.
-
-It would be very difficult to implement this pattern without direct support from WebAssembly because two interface methods assigned to the same slot can have completely different signatures, i.e. number and size of arguments.
-So call tags enables an important pattern used in practice to support a feature that is critical for many major languages (specifically Java, Kotlin, and Scala come to mind, though not C# due to its decision to support multiple-instantiation inheritance of generic interfaces).
-
-#### Dynamic Arity
+#### Representation Specialization vs Signature Refinement
+
+Now consider a class like `class Super { void bar(Int i) {...} }` in Dart or Kotlin.
+One would like to use unboxed ints when calling this method.
+But, due to contravariance of parameter types, it's possible that there's a subclass `class Sub { void bar(Object o) {...} }` refining the type of `bar` to take an arbitrary object.
+This subclass needs the argument to be boxed into the uniform representation.
+So, especially in the presence of incremental rather than whole-program compilation, does the possibility of that subclass mean that `Super.bar` needs to implemented boxed ints?
+
+We can solve this problem using call tags, even in the presence incremental compilation.
+We again generate two call tags, but this time we use the most natural specialized representation for the signature at hand.
+This means we have both `$Super.bar : [i32] -> []` and `$Sub.bar : [(ref $Object)] -> []`.
+Then, rather `Sub` using a single `func` to implement both call tags for `bar`, we have `Sub` use a `func_switch`.
+For `$Super.bar`, it simply switches to `Sub`'s implementation of `bar`.
```

Comment:
Are these reversed or something? For Super I would expect it to box the int and for Sub to pass through.